x/ibc: testing pkg

colin-axner opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description
Adding `TestChain` and a `Coordinator` to new pkg `x/ibc/testing`. Will support general IBC functionality via message execution as opposed to previous method of direct keeper updates.

first part of split from: #6329 

ref: #5558 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

colin-axner commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,298 @@
+	// increment the current header
+	chain.CurrentHeader = abci.Header{
+		Height: chain.CurrentHeader.Height + 1,
+		Time:   time.Now(),
```
thoughts on using `Now()` in testing? I'm not aware of how common non-deterministic bugs arise from this usage. 

I'm thinking the more robust approach is to add chain func to take in a counterparty chain and update the chains time and the counterparty's time by 5 seconds (or something like that). This limits this testing pkg to be used for 2 chains at a time only. 

This may be easier to implement if I allow references from this TestChain to the testing suite holding all the test chains

AdityaSripal replied:
Yea I like just taking the previous time and updating by like 5 seconds

AdityaSripal commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,298 @@
+
+// UpdateClient will construct and execute a 07-tendermint MsgUpdateClient. The counterparty
+// client will be updated on the (target) chain.
+func (chain *TestChain) UpdateClient(counterparty *TestChain) {
```
Huge improvements!! Awesome that we're doing this instead of the previous approach

AdityaSripal commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,298 @@
+
+// NewCapability creates a new capability for the provided port and channel ID
+// if the capability does not already exist.
+func (chain *TestChain) NewCapability(portID, channelID string) (*capabilitytypes.Capability, error) {
```
Why would we ever need this function directly? Shouldn't this be handled by registering ports, creating channels, etc

colin-axner replied:
yup, leftover from the old way of handling create conn/channel, will be removed once those are updated.

colin-axner commented:
I added a `suite.go` file which contains `IBCTestSuite` this struct will be used for interfacing with chains and keeping them in sync. 

The basic idea is developers will init a testsuite with the number of chains they need and then refer to the chains by indexes using the test suite funcs. 

I think some readability is sacrificed with this approach, but I'm not sure of any other way of preventing the chains from getting out of sync other than managing them at some middle level in between the chain (lowest) and module (highest)

I will update the create conn/channel funcs next. Let me know if there is better approach

colin-axner commented on x/ibc/testing/suite.go:
```diff
@@ -0,0 +1,179 @@
+
+// IBCTestSuite is a testing struct which wraps the testing suite with N TestChain's. It handles
+// keeping all chains in time-sync.
+type IBCTestSuite struct {
```
I'm not sure if this is the best naming. Maybe it should be renamed and just receive a `suite.T()` for testing in its constructor

colin-axner replied:
I update it to `Coordinator`, still unsure about the naming, but feels more accurate than before

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+type TestChain struct {
+	t *testing.T
+
+	ClientID      string
+	App           *simapp.SimApp
+	LastHeader    ibctmtypes.Header // header for last block height committed
+	CurrentHeader abci.Header       // header for current block height
+	Querier       sdk.Querier
+
+	Vals    *tmtypes.ValidatorSet
+	Signers []tmtypes.PrivValidator
+
+	senderPrivKey crypto.PrivKey
+	SenderAccount authtypes.AccountI
+
+	// IBC specific helpers
+	Channels    []Channel // track portID/channelID's created for this chain
+	Connections []string  // track connectionID's created for this chain
```
```suggestion
	App           *simapp.SimApp
	LastHeader    ibctmtypes.Header // header for last block height committed
	CurrentHeader abci.Header       // header for current block height
	Querier       sdk.Querier

	Vals    *tmtypes.ValidatorSet
	Signers []tmtypes.PrivValidator

	senderPrivKey crypto.PrivKey
	SenderAccount authtypes.AccountI

	// IBC specific helpers
	ClientID      string
	Connections []string  // track connectionID's created for this chain
	Channels    []Channel // track portID/channelID's created for this chain
```

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+	ConnectionVersion = "1.0"
+	ChannelVersion    = "1.0"
+
+	ConnectionIDPrefix = "connectionID"
+	ChannelIDPrefix    = "channelID"
+	PortIDPrefix       = "portID"
```
what are these for?

colin-axner replied:
naming prefixes when creating connections and channels. This is just a substitution to `testPort1` etc

colin-axner replied:
alternatively these could be randomly generated, but I figured something readable is useful ie `channel/channelID0/port/portID0`
`channel/channelID1/port/portID1`

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+// at the next block height. It does not update the time as that is handled by the IBCTestSuite.
+//
+// CONTRACT: this function must only be called after app.Commit() occurs
+func nextBlock(chain *TestChain) {
```
I'd make this public so that other SDK apps can use it:

```suggestion
func (chain *TestChain) NextBlock() {
```

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+	require.NoError(chain.t, err)
+
+	// SignCheckDeliver calls app.Commit()
+	nextBlock(chain)
```
```suggestion
	chain.NextBlock()
```

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+
+// CreateClient will construct and execute a 07-tendermint MsgCreateClient. A counterparty
+// client will be created on the (target) chain.
+func (chain *TestChain) CreateClient(counterparty *TestChain) {
```
we should add a `clientType` field to create other client types via `switch`

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+type TestChain struct {
+	t *testing.T
+
+	ClientID      string
```
should this be an array? in case a chain has both a localhost and tendermint clients defined for example? Or would one rather create a new `TestChain` instance for that?

colin-axner replied:
I think it makes sense to test having multiple clients on a chain. Do we want to autogenerate clientID names like done for connections/channels/ports?

fedekunze replied:
> Do we want to autogenerate clientID names like done for connections/channels/ports?

Yeah that makes sense 👍

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+	SenderAccount authtypes.AccountI
+
+	// IBC specific helpers
+	Channels    []Channel // track portID/channelID's created for this chain
```
I don't fully understand what these helpers are for. Are they for ease of retrieving the channels and connections? ~I don't see that they are being updated thought the handshakes. Should we use pointers instead or a `map [portID+channelID] -> *Channel`? Thoughts~

fedekunze replied:
ahh lol, I thought you were using the actual channel struct. Ignore the second part of the comment

colin-axner replied:
yea I was worried that naming might be a bit confusing. Maybe I should rename it to `TestChannel` or something to be more explicit that it is quite different an actual IBC channel

fedekunze replied:
`TestChannel` is better

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+
+// UpdateClient will construct and execute a 07-tendermint MsgUpdateClient. The counterparty
+// client will be updated on the (target) chain.
+func (chain *TestChain) UpdateClient(counterparty *TestChain) {
```
ditto `clientType`

fedekunze commented on x/ibc/testing/chain.go:
```diff
@@ -0,0 +1,373 @@
+	// construct MsgCreateClient using counterparty
+	msg := ibctmtypes.NewMsgCreateClient(
+		counterparty.ClientID, counterparty.LastHeader,
+		lite.DefaultTrustLevel, TrustingPeriod, UnbondingPeriod, MaxClockDrift,
```
`lite.DefaultTrustLevel` should also be a variable defined at the top level of the file so that it can be overwritten in a case it's necessary

fedekunze commented on x/ibc/testing/channel.go:
```diff
@@ -0,0 +1,26 @@
+	}
+}
+
+// GetPortID returns the PortID.
+func (ch Channel) GetPortID() string {
+	return ch.PortID
+}
+
+// GetChannelID returns the ChannelID.
+func (ch Channel) GetChannelID() string {
+	return ch.ChannelID
+}
```
There's no need for these funcs imho since the fields are public

fedekunze commented on x/ibc/testing/coordinator.go:
```diff
@@ -0,0 +1,178 @@
+// Coordinator is a testing struct which contains N TestChain's. It handles keeping all chains
+// in sync with regards to time.
+type Coordinator struct {
+	Chains []*TestChain
```
I'd also add a priv field to track the index in the array from a given ChainID: `map[string]int` (i.e `map[chainID]idx`). Then you can use chainIDs for each of the functions which seems more intuitive to me.

colin-axner replied:
I agree that indexes are not the best to work with. Maybe just having a `map[string]*TestChain`?

colin-axner replied:
I think turning the `ClientID` field to `ChainID` (and making a client id array) makes sense with respect to this mapping.

fedekunze commented on x/ibc/testing/coordinator.go:
```diff
@@ -0,0 +1,178 @@
+	source := coord.Chains[sourceIndex]
+	counterparty := coord.Chains[counterpartyIndex]
+
+	if state == connectiontypes.UNINITIALIZED {
+		return
+	}
```
imo we should remove these checks so that if the handshake fails we know that something is misconfigured or we performed the wrong steps during the handshake

colin-axner replied:
So my main motivation for adding these was allowing the caller to initialize a conn/channel with an intermediate state while keeping the code DRY. I guess since using intermediate states is less likely we could push that onto the developer or add funcs that only execute the intermediate state despite the redundant code

AdityaSripal replied:
Yea makes sense to have these returns so we can return partially created connections to our test functions

fedekunze commented on x/ibc/testing/coordinator.go:
```diff
@@ -0,0 +1,178 @@
+
+	// Initialize connection on source
+	source.ConnectionOpenInit(counterparty, sourceConnection, counterpartyConnection)
+	coord.IncrementTime()
```
should we update the client too?

colin-axner replied:
oh yea definitely

AdityaSripal commented on x/ibc/testing/coordinator.go:
```diff
@@ -0,0 +1,178 @@
+	// Open connection on both chains
+	source.ConnectionOpenAck(counterparty, sourceConnection, counterpartyConnection)
+	coord.IncrementTime()
+
```
Shouldn't there be this check here

```suggestion
if state == connectiontypes.ACK {
    return
}
```

colin-axner replied:
that type DNE, see [connection states](https://github.com/cosmos/cosmos-sdk/blob/08b7557fe5512cf5a0a0f8143e6b21c4d2950922/x/ibc/03-connection/types/types.pb.go#L32)

AdityaSripal commented on x/ibc/testing/coordinator.go:
```diff
@@ -0,0 +1,178 @@
+//
+// NOTE: The counterparty testing connection will be created even if it is not created in the
+// application state.
+func (coord *Coordinator) CreateConnection(
```
Like how this function is constructed. Though I think we should allow connection to end at any CONNECTION STATE (including ACK). I also think we should specify the counterparty Connection state I want to end at.

It may also be useful to allow this function to try faulty configs: (i.e. STATE: INIT, COUNTERPARTY: CONFIRM)

This function tries to alternate handshake functions and returns error if one of the handshake msgs do.

This could test if incorrect handshakes return error. But understood, if you don't really want to do that in here

colin-axner commented:
I think all the stuff mentioned above has been addressed. I added a `TestConnection` struct for convenience and moved it to `types.go` with `TestChannel`.

Currently updating a client in create connection and channel is hard coded to use a TM client. I think this should probably be updated in the future to be retrieved from using the ClientID info, but I think this is best done in a separate pr given this one has already grown quite a bit. 

Some handling here I think can be simplified and var naming within `CreateConnection` and `CreateChannel` should probably be modified since over usage of `source` and `counterparty` can be confusing, but unless someone has a suggestion rn I think it can wait. I figure this code will need to be modified as the tests are reworked to use the testing pkg anyways.

fedekunze commented on x/ibc/testing/coordinator.go:
```diff
@@ -50,84 +53,210 @@ func (coord *Coordinator) IncrementTime() {
-		chain := coord.Chains[index]
+func (coord *Coordinator) CommitBlock(chains ...string) {
+	for _, chainID := range chains {
+		chain := coord.Chains[chainID]
```
check for error, applies for all the `coord.Chains[chainID]`

```suggestion
		chain, ok := coord.Chains[chainID]
		require.True(t, ok, fmt.Sprintf("%s chain doesn't exist", chainID))
```

fedekunze replied:
you can wrap it on a function too 👍
